test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
+                  const af::dtype ty) {
+    const dim_t pad_size = 2;
+
+    // The large array is padded on both sides of each dimension
+    // Padding is only applied if the dimension is used, i.e. if dims[i] > 1
+    af::dim4 full_arr_dims(dims[0] > 1 ? dims[0] + 2*pad_size : dims[0],
+                           dims[1] > 1 ? dims[1] + 2*pad_size : dims[1],
+                           dims[2] > 1 ? dims[2] + 2*pad_size : dims[2],
+                           dims[3] > 1 ? dims[3] + 2*pad_size : dims[3]);
+
+    // Calculate index of sub-array. These will be used also by
+    // testWriteToOutputArray so that the gold sub array will be placed in the
+    // same location. Currently, this location is the center of the large array
+    af::index subarr_idxs[4];
+    for (int i = 0; i < 4; ++i) {
+        subarr_idxs[i] =
+            dims[i] > 1 ? af::seq(pad_size, pad_size + dims[i] - 1) : af::span;
+    }
+
+    metadata.initArrays(full_arr_dims, ty, subarr_idxs);
+}
+
+// Generates a reordered array. testWriteToOutputArray expects that this array
+// will still have the correct output values from the af_* function, even though
+// the array was initially reordered.
+void genReorderedArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    // This reorder combination will not move data around, but will simply
+    // call modDims and modStrides (see src/api/c/reorder.cpp).
+    // The output will be checked if it is still correct even with the
+    // modified dims and strides "hack" with no data movement
+    unsigned reorder_0 = 0;
+    unsigned reorder_1 = 2;
+    unsigned reorder_2 = 1;
+
+    // Shape the output array such that the reordered output array will have
+    // the correct dimensions that the test asks for (i.e. must match dims arg)
+    af::dim4 out_dims(dims[reorder_0], dims[reorder_1], dims[reorder_2]);
+    metadata.initArrays(out_dims, ty);
+    metadata.setOutputArray(af::reorder(metadata.out_arr,
+                                        reorder_0, reorder_1, reorder_2));
+}
+
+// Partner function of testWriteToOutputArray. This generates the "special"
+// array that testWriteToOutputArray will use to check if the af_* function
+// correctly uses an existing array as its output
+void genTestOutputArray(af_array *out_ptr,
+                        const unsigned ndims, const dim_t *dims,
+                        const af::dtype ty,
+                        TestOutputArrayInfo* metadata) {
+    af::dim4 arr_dims(ndims, dims);
+    switch (metadata->getOutputArrayType()) {
+    case NULL_ARRAY:
+        genNullArray(*metadata, arr_dims, ty);
+        break;
+    case FULL_ARRAY:
+        genRegularArray(*metadata, arr_dims, ty);
+        break;
+    case SUB_ARRAY:
+        genSubArray(*metadata, arr_dims, ty);
+        break;
+    case REORDERED_ARRAY:
+        genReorderedArray(*metadata, arr_dims, ty);
+        break;
+    }
+    *out_ptr = metadata->getOutputArrayC();
```

Comment:
will become
```C++
af_array out = 0;
if (metadata->getOutputArrayType() == SUB_ARRAY) {
    af::array original = metadata->getOutputArray();
    af::array subarr = original(metadata->idx(0), metadata->idx(1),
                                metadata->idx(2), metadata->idx(3));
    // may be it will be better to use C-API in above statement
    af_retain_array(&out, subarr);
} else {
    af_retain_array(&out, metadata->getOutputArray().get());
}
*out_ptr = out;
```

Reply:
A couple of concerns here:
- I've written about storing the original, full array and making a copy of it for the actual testing later. Please see my comment on `genSubArray`.
- Maybe it is better indeed to use the C API there. Either way though we'll need to retain `subarr` or `metadata->getOutputArray().get()`. The question is where in the code do we eventually release it. My solution to this (formed from Umar's suggestion to have the destructor call the release) is to store the value of `af_retain_array`'s output in `TestOutputArrayInfo`, then eventually release that inside the destructor. This essentially makes a central point on where to release whatever was allocated/retained, regardless of whether the asserts pass or fail (since the destructor is called either way).